### PR TITLE
CurrentUserOnly should allow multiple servers on Windows

### LINKED ISSUE #28371: NamedPipeServerStream throws UnauthorizedException when creating multiple server streams on windows with PipeOptions.CurrentUserOnly
Repro:
```c#
static void Main(string[] args)
{
    const int PipeBufferSize = 0x10000;

    var pipeName = Guid.NewGuid().ToString();
    // This will throw an exception after the first loop. Remove "PipeOptions.CurrentUserOnly" to make this work.
    for (var i = 0; i < 5; i++) {
        var server = new NamedPipeServerStream(
            pipeName,
            PipeDirection.InOut,
            NamedPipeServerStream.MaxAllowedServerInstances,
            PipeTransmissionMode.Byte,
            PipeOptions.Asynchronous | PipeOptions.CurrentUserOnly | PipeOptions.WriteThrough,
            PipeBufferSize,
            PipeBufferSize);
        using (var client = new NamedPipeClientStream(".", pipeName, PipeDirection.InOut, PipeOptions.Asynchronous | PipeOptions.CurrentUserOnly))
        {
            client.Connect();
            Console.WriteLine("Connected!");
        }
    }
}
```

cc @pjanotti @rynowak

### PR DESCRIPTION
**Pull request by pjanotti**
Fix #28371

### CODE REVIEW

**Comment by pjanotti** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build
@dotnet-bot test Outerloop OSX x64 Debug Build

File: src/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.CurrentUserOnly.netcoreapp.cs [hunk 1]
```diff
@@ -116,5 +116,37 @@ public static void CreateMultipleServers_ConnectMultipleClients_MultipleThreads(
 
             Task.WaitAll(tasks.ToArray());
         }
+
+        [Fact]
+        public static void CreateMultipleConcurrentServers_ConnectMultipleClients()
```
Comment (stephentoub) [inline src/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.CurrentUserOnly.netcoreapp.cs]: Does this test work on unix?

Comment (pjanotti) [reply src/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.CurrentUserOnly.netcoreapp.cs]: Nope...

**Comment by pjanotti** [comment]
@safern without CurrentUserOnly can be valuable to enforce current behavior. The negative test w/ multiple servers seems redundant: the DACL should prevent the client to connect to the pipe, there is no reaction on the Windows pipe server in this case.

File: src/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.CurrentUserOnly.netcoreapp.cs (hunk 1 above)
Comment (pjanotti) [reply src/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.CurrentUserOnly.netcoreapp.cs]: It is failing on a Debug.Assert inside SharedServer.Get: it wants max number of services to be positive, however, we are keeping the sentinel value -1 (MaxAllowedServerInstances). So it is working as intended but we either change the assert or the max number of services to int.MaxValue if the sentinel value was found. I will update the PR.

**Comment by pjanotti** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build
@dotnet-bot test Outerloop OSX x64 Debug Build

**Comment by pjanotti** [comment]
Test failures are not related to the changes.

**Pull request merged**